docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,189 @@
+Nullable Metadata
+=========
+The following describes the representation of nullable annotations in metadata.
+
+## NullableAttribute
+Type references are annotated in metadata with a `NullableAttribute`.
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(
+        AttributeTargets.Class |
+        AttributeTargets.Event |
+        AttributeTargets.Field |
+        AttributeTargets.GenericParameter |
+        AttributeTargets.Parameter |
+        AttributeTargets.Property |
+        AttributeTargets.ReturnValue,
+        AllowMultiple = false)]
+    public sealed class NullableAttribute : Attribute
+    {
+        public readonly byte[] Flags;
+        public NullableAttribute(byte flag)
+        {
+            Flags = new byte[] { flag };
+        }
+        public NullableAttribute(byte[] flags)
+        {
+            Flags = flags;
+        }
+    }
+}
+```
+
+The `NullableAttribute` type is for compiler use only - it is not permitted in source.
+The type declaration is synthesized by the compiler if not already included in the compilation.
+
+Each type reference in metadata may have an associated `NullableAttribute` with a `byte[]` where each `byte`
+represents nullability: 0 for oblivious, 1 for not annotated, and 2 for annotated.
+
+The `byte[]` is constructed as follows:
+- Reference type: the nullability (0, 1, or 2), followed by the representation of the type arguments in order including containing types
+- Nullable value type: the representation of the type argument only
+- Value type: the representation of the type arguments in order including containing types
+- Array: the nullability (0, 1, or 2), followed by the representation of the element type
+- Tuple: the representation of the tuple elements in order
+- Type parameter reference: the nullability (0, 1, or 2, with 0 for unconstrained type parameter)
+
+Note that certain type references are represented by an empty `byte[]`:
+- Non-generic value types, and
+- Nullable value types and tuples where all elements are represented by an empty `byte[]`.
+
+### Optimizations
+
+If the `byte[]` is empty, the `NullableAttribute` is omitted.
+
+If all values in the `byte[]` are the same, the `NullableAttribute` is constructed with that single `byte` value. (For instance, `NullableAttribute(1)` rather than `NullableAttribute(new byte[] { 1, 1 }))`.)
+
+### Type parameters
+Each type parameter definition may have an associated `NullableAttribute` with a single `byte`:
+
+1. `notnull` constraint: `NullableAttribute(1)`
+2. `class` constraint in `#nullable disable` context: `NullableAttribute(0)`
+3. `class` constraint in `#nullable enable` context: `NullableAttribute(1)`
+4. `class?` constraint: `NullableAttribute(2)`
+5. No `notnull`, `class`, `struct`, `unmanaged`, or type constraints in `#nullable disable` context: `NullableAttribute(0)`
```

Comment:
>`NullableAttribute(0) [](start = 101, length = 21)

Today we never emit such an attribute or specially recognize it on import. Are you proposing a change?